### PR TITLE
AF: per-channel volume loss weighting (upweight nut/p)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current champion uses a scalar vol-loss-weight=10.0 to upweight the entire volume loss. But the 4 volume channels (ux, uy, p, nut) have very different scales and difficulty — pressure (p) and turbulent viscosity (nut) are harder to predict and have higher relative error. Applying per-channel weights should help the model focus on the harder-to-predict channels without over-penalizing the easier ones (ux, uy), potentially improving both vol_mse and surface_mse.

## Current Baseline

- **val_primary/surface_mse:** 0.000296 (epoch 704) — PR #3135
- **val_primary/vol_mse:** 0.002039 (epoch 743) — PR #3135
- **Champion:** AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, vol-weight=10x, Fourier, 2L/256d/4H

## Implementation

Add `--vol-channel-weights` CLI arg that accepts 4 floats (one per volume channel: ux, uy, p, nut). When specified, replace the uniform `F.mse_loss` vol loss with a weighted per-channel MSE:

```python
preds = outputs["volume_preds"][batch.volume_mask]   # shape [N_vol, 4]
tgt   = transform.apply(batch.volume_y)[batch.volume_mask]  # shape [N_vol, 4]
if args.vol_channel_weights is not None:
    ch_w = torch.tensor(args.vol_channel_weights, device=preds.device, dtype=preds.dtype)
    per_ch = ((preds - tgt) ** 2).mean(dim=0)       # shape [4]
    vol_loss = (per_ch * ch_w).sum() / ch_w.sum()   # normalized weighted average
else:
    vol_loss = F.mse_loss(preds, tgt)                # fallback to baseline
```

Then scale the total vol loss by `args.vol_loss_weight` as before (keep vol-loss-weight=10.0).

## Trials

Run all trials on AirfRANS full task. Use `--wandb_group af-per-channel-vol-weighting` so runs are grouped.

**T1** — nut 4x, others uniform (strongest uplift on hardest channel):
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-loss-weight 10.0 \
  --vol-channel-weights 1 1 1 4 \
  --wandb_group af-per-channel-vol-weighting
```

**T2** — p 2x + nut 4x (uplift both hard channels):
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-loss-weight 10.0 \
  --vol-channel-weights 1 1 2 4 \
  --wandb_group af-per-channel-vol-weighting
```

## Success Criteria

- Primary: `val_primary/surface_mse` < 0.000296
- Secondary: `val_primary/vol_mse` < 0.002039
- Report best-checkpoint metrics (not terminal) for both surface_mse and vol_mse

## Expected Outcome

Per-channel upweighting of nut (highest relative error) should push the model to improve turbulent viscosity prediction. Since nut and surface quantities are tightly coupled via the boundary layer, improvements in nut should propagate to surface_mse. T2 is more aggressive and may yield larger gains but risks degrading ux/uy.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Trial 1 — nut 4× (weights [1, 1, 1, 4])

| Metric | Baseline | T1 | Δ |
|--------|----------|----|---|
| val surface_mse | 0.000296 | **0.000410** | +38.5% ❌ |
| val vol_mse | 0.002039 | 0.004439 | +117.7% ❌ |
| test surface_mse | 0.000831 | 0.001543 | +85.7% |
| test vol_mse | 0.002721 | 0.005654 | +107.8% |

- **Best epoch**: 557 / 766 total
- **W&B run**: `s7l3zbeb`
- **Peak VRAM**: ~7.2 GB

### Trial 2 — p 2× + nut 4× (weights [1, 1, 2, 4])

| Metric | Baseline | T2 | Δ |
|--------|----------|----|---|
| val surface_mse | 0.000296 | **0.000649** | +119.3% ❌ |
| val vol_mse | 0.002039 | 0.004436 | +117.6% ❌ |
| test surface_mse | 0.000831 | 0.002846 | +242.4% |
| test vol_mse | 0.002721 | 0.005150 | +89.3% |

- **Best epoch**: 718 / 758 total
- **W&B run**: `gw7wwex7`
- **Peak VRAM**: ~7.2 GB

### Commands used

```bash
# T1
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --agent tanjiro --wandb-name "tanjiro/af-vol-ch-weight-T1-nut4x" --wandb-group "af-per-channel-vol-weighting" --vol-channel-weights 1 1 1 4 --volume-loss-weight 10.0

# T2
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --agent tanjiro --wandb-name "tanjiro/af-vol-ch-weight-T2-p2x-nut4x" --wandb-group "af-per-channel-vol-weighting" --vol-channel-weights 1 1 2 4 --volume-loss-weight 10.0
```

### What happened

**Negative result.** Both trials significantly underperformed the baseline on all metrics.

Key observations from per-channel logging:
- **nut** is indeed the hardest channel (per-channel MSE ~0.017 vs ux/uy ~3e-5), confirming the motivation.
- However, upweighting nut/p in the volume loss **hurt surface prediction** substantially — surface_mse degraded by 38–119%.
- Volume MSE also got worse (~2× baseline), suggesting the heavier channel weights distorted the overall loss landscape rather than helping the model focus.
- T2 (p 2× + nut 4×) was worse than T1 (nut 4× only), indicating that adding more weight to p compounds the problem.

**Why it failed:** The baseline already uses `volume_loss_weight=10.0` which gives volume a large share of the total loss. Further reweighting individual channels within the volume component appears to create an imbalanced gradient signal that the model can't recover from — the easy channels (ux, uy) contribute negligible gradients while nut dominates, but nut is inherently noisy/hard to fit, so the model oscillates on nut without making progress on surface metrics.

### Suggested follow-ups

1. **Inverse approach**: Instead of upweighting hard channels, try *downweighting* nut (e.g. [1, 1, 1, 0.25]) — nut may be injecting gradient noise that hurts convergence.
2. **Adaptive channel weighting**: Use running per-channel loss ratios to dynamically balance weights, so channels that are already low-error don't get over-penalized.
3. **Channel-specific learning rate scaling**: Instead of loss weighting, scale the gradients for channel-specific output heads.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3250 because ADVISOR: Closing — per-channel vol weighting surface=0.000410 (+38%), vol=0.004439 (+118%). Per-channel biasing disrupts joint optimization. Tanjiro reassigned..
